packages/node/src/indexer/ds-plugin.service.ts

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020-2021 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import assert from 'assert';
+import path from 'path';
+import { Injectable } from '@nestjs/common';
+import {
+  SubqlCustomDatasource,
+  SubqlDatasourcePlugin,
+  SubqlNetworkFilter,
+} from '@subql/types';
+import { NodeVM, NodeVMOptions, VMScript } from '@subql/x-vm2';
+import { merge } from 'lodash';
+import { SubqueryProject } from '../configure/project.model';
+import { getLogger } from '../utils/logger';
+import { isCustomDs } from '../utils/project';
+
+export interface DsPluginSandboxOption {
+  root: string;
+  entry: string;
+}
+
+const DEFAULT_OPTION: NodeVMOptions = {
+  console: 'redirect',
+  wasm: false,
+  sandbox: {},
+  require: {
+    builtin: ['assert', 'buffer', 'crypto', 'util', 'path'],
+    external: true,
+    context: 'sandbox',
+  },
+  wrapper: 'commonjs',
+  sourceExtensions: ['js', 'cjs'],
+};
+
+const logger = getLogger('ds-sandbox');
+
+export class DsPluginSandbox extends NodeVM {
+  private option: DsPluginSandboxOption;
+  private script: VMScript;
+  private entry: string;
+
+  constructor(option: DsPluginSandboxOption) {
+    const vmOption: NodeVMOptions = merge({}, DEFAULT_OPTION, {
+      require: {
+        root: option.root,
+        resolve: (moduleName) => {
+          return require.resolve(moduleName, { paths: [option.root] });
+        },
+      },
+    });
+    super(vmOption);
+    this.option = option;
+    this.entry = option.entry;
+    this.script = new VMScript(
+      `
+      module.exports = require('${this.entry}');
+    `,
+      path.join(option.root, 'ds_sandbox'),
+    );
+  }
+
+  getDsPlugin<
+    D extends string,
+    T extends SubqlNetworkFilter,
+  >(): SubqlDatasourcePlugin<D, T> {
+    return this.run(this.script);
+  }
+}
+
+@Injectable()
+export class DsPluginService {
```

Comment:
It would be cool to use this for data source VMs as well, there's a lot of overlap

Reply:
> I think this is a good approach, I assume that support for Ink would be very similar.
> 
> I have a few questions though:
> 
>     * How do you see ds-plugins being distributed? Would each project reference them?
> 
>     * Can this be used to potentially support other blockchains that aren't substrate based? And how would that effect this approach if we wanted to support other blockchains

For 1), we will publish package to support acala/moonbeam/ink, user include our package in package.json and refer via 
```
datasources:
   - processor:
        file: node_modules/@subql/contract-processors/ink.js
```
or they can write their own
```
datasources:
   - processor:
        file: dist/own-processor.js
```
For 2), I don't think this can support other protocols. It was built on an assumption that these contract calls and events are encapsulated in substrate extrinsic and event.